contracts/MarginCalculator.sol

```diff
@@ -43,19 +36,227 @@ contract MarginCalculator {
+    uint256 internal constant BASE = 8;
+
+    /// @dev oracle deviation value (1e27)
+    uint256 internal oracleDeviation;
+
+    /// @dev FixedPoint 0
+    FPI.FixedPointInt internal ZERO = FPI.fromScaledUint(0, BASE);
+
+    /// @dev mapping to store dust amount per option collateral asset (1e27)
+    mapping(address => uint256) internal dust;
+
+    /// @dev mapping to store array of time to expiry per product
+    mapping(bytes32 => uint256[]) internal productTimeToExpiry;
+
+    /// @dev mapping to store option upper bound value at specific time to expiry per product (1e27)
+    mapping(bytes32 => mapping(uint256 => uint256)) internal timeToExpiryValue;
+
+    /// @dev mapping to store shock value for spot price per product (1e27)
+    mapping(bytes32 => uint256) internal spotShock;
+
+    /// @notice emits an event when collateral dust is updated
+    event CollateralDustUpdated(address indexed collateral, uint256 dust);
+    /// @notice emits an event when new time to expiry is added for a specific product
+    event ProductTimeToExpiryAdded(bytes32 indexed productHash, uint256 timeToExpiry);
+    /// @notice emits an event when new upper bound value is added for a specific time to expiry timestamp
+    event TimeToExpiryValueAdded(bytes32 indexed productHash, uint256 timeToExpiry, uint256 value);
+    /// @notice emits an event when spot shock value is updated for a specific product
+    event SpotShockUpdated(bytes32 indexed product, uint256 spotShock);
+
     constructor(address _oracle) public {
         require(_oracle != address(0), "MarginCalculator: invalid oracle address");
 
         oracle = OracleInterface(_oracle);
     }
 
+    /**
+     * @notice set dust amount for collateral asset (1e27)
+     * @dev can only be called by owner
+     * @param _collateral collateral asset address
+     * @param _dust dust amount
+     */
+    function setCollateralDust(address _collateral, uint256 _dust) external onlyOwner {
+        dust[_collateral] = _dust;
+    }
+
+    /**
+     * @notice set new time to expiry for specific product
+     * @dev can only be called by owner
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _timeToExpiry option time to expiry timestamp
+     */
+    function setProductTimeToExpiry(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _timeToExpiry
+    ) external onlyOwner {
+        bytes32 productHash = keccak256(abi.encode(_underlying, _strike, _collateral, _isPut));
+        uint256[] storage expiryArray = productTimeToExpiry[productHash];
+
+        require(
+            (expiryArray.length == 0) || (_timeToExpiry > expiryArray[expiryArray.length.sub(1)]),
+            "MarginCalculator: expiry array is not in order"
+        );
```

Comment:
so we can't update any timestamp in the the middle of the expiryArray array?

Reply:
Yes